### ISSUE TITLE
Phoenix Synergy 09/28/2019 NCSM 1st Cycle Accounts

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1738]
In GitLab by @kdjstudios on Oct 10, 2019, 09:09

**Submitted by:** Cheryl Hamelin <cheryl.hamelin@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/9485839

**Server:** Internal

**Client/Site:** Phoenix Synergy

**Account:** NA

**Issue:** 

Regarding the Phoenix Synergy 09/28/2019 NCSM 1st Cycle Accounts, Julian and I created the billing cycle for this account. When we tried to put in the amounts for the invoice, it shows that the invoice will have a 1/1/19 date. If you look in the staged fee section, it says the cycle is for the 12/28/18 cycle.

I have attached the screen shot for your review. Please correct this so that I may finish my billing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 09:16

I wrote:

> I just took a quick look at the Phoenix synergy NCSM billing cycle and do not see anywhere it mentions it will have an 1/1/19 invoice date? Everywhere I show the invoice date it states 10/1/19 which appears to be correct. Please note the billing cycle name and date are not the same as the invoice date. The invoice date is calculated based upon the setup of the Billing cycle on the site edit page.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 09:42

Jualian wrote:

> The Draft invoice for Phoenix Synergy/Plane Source/130-N0037/invoice# 130-500 - still shows "invoice date" as 01/01/2019. 

@amishra @smasih 

What is going on? This appears this is another issue with Billing cycles?

I ran the Review draft invoice report and saw that this report shows that account had it's invoice at a different time? Also I presume that the report is showing the "create at date" and not the "invoice date" in the invoice date column? Please confirm as we may need to fix the report.

[draft_invoice_detail_10_10_2019.pdf](/uploads/6d45e26856bedbca83b6e63a9f1af74e/draft_invoice_detail_10_10_2019.pdf)

Based on the assumption that is the created at date. How did the site create a billing cycle invoices the following day? Should there not be a check for existing draft invoices before they can have a billing cycle create draft invoices?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 10:01

I wrote:

> Thank you for the update. From a quick look at that account and comparing with other accounts on the same site; it would appear that the invoice on account 130N0037 was generated  before the other invoices and not as part of the billing cycle, but was manually created by a user. We will have the support team look further into this and follow up with their findings. 

Cheryl wrote:

> When I sent my original  request to you, I included screen shots that were sent as an attachment and showed exactly what I was looking at. Maybe you did not see the attachment. If you go back to the original email, everything is there including the exact account.
> 
> Thanks for looking into this for us.

I wrote:

> I also just realized that the 9/28/19 Billing cycle has already been finalized. Which supports my conclusion below that the invoice on that particular account was created manually and not part of the billing cycle, thus why it's invoice date did not match. If that invoice was part of the billing cycle it would have been finalized when the billing cycle was finalized. 

I wrote:

> Thank you, and we will need to have the support team look into the logs, as I said it would appear a user has been making manual adjustments and did not let the Billing cycle complete as it should have. Once the support team is able to look at the logs we will follow up with their findings.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 10:07

@smasih @amishra 

I also pulled the Billing cycle page and it would appear that the site has already finazlied the Billing cycle. So this would lead me to believe that the particular draft invoice they are mentioning was not part of the billing cycle somehow? but how did the screenshot she sent over show that those pending staged fees were supposed to be?

Also just found out that the link on that billing cycle page is not working!!! What is going on!!! LINK: https://answernet.sabilling.com/reports/InvoicesBySite?utf8=%E2%9C%93&site_id=5b59e747cf85c1318b000097&billing_cycle_id=5d9e4151cf85c162e8001239&commit=GO&format=html

![image](/uploads/61de505968b500eb9d7e804042075f49/image.png)

So I manually ran the report for that billing cycle, which shows it should have 5 invoices. But now it shows there are no invoices?????!!!!

![image](/uploads/9358033bb9bb27cb2df94f471013f9ff/image.png)

I also tried to just run the Invoice by site report, and it does not even load!!

![image](/uploads/7ea01a198e29ae52d9d578f5c26ab2ff/image.png)


Escalating this to Whatsapp

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 10:24

Cheryl wrote:

> The invoice was created before we even started the billing. For that site. Julian put it in as staged fees before we even started. I do not know why it is finalized because I was here late with Michelle McKee last night and we were not finalizing it because I wanted to hear back from you as to how to fix the issue before finalizing. If I log out of the system and leave it open, would that make it finalize without me doing it on my own?
> 
> Yesterday, Julian and I went through the sites and had made changes and finalized Sarasota, but when I came in this morning, it was not finalized and when I checked the Billing Cycle report, the changes Julian and I made were not there and I had to make them again and then finalize. 
> 
> I have been having other people look at things that I find so that they see the same thing I do in order to verify what I am finding.

I wrote:


> The support team will need to check the logs to see which user(s) performed the finalization and determine the other steps that occurred to cause this discrepancy as I do not have access to that. 
> 
> As long as no one has sent out invoices for this billing cycle, I would propose the following solution.
> 
> 1. Revert the billing cycle.
> 2. Delete any manually drafted invoices that still remain.
> 3. Confirm all staged fees are added and are showing they will be added to the correct billing cycle.
> 4. Reprocess your billing cycle.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 12:37

@smasih @jredfield 

I informed the site to perform a revert and start over in hopes that will temporary resolve the issue, since we were not getting updates here. once the site is able to reprocess the billing cycle if they continue to have issues, we will post those in the urgent items here. 

Please look into the following Urgent items:

1. Why the invoice by site report is not able to load.

The following non urgent items can be looked at tomorrow. 

1. Look into the Review draft invoices report and determine why the end date range keeps incrementing every time we run the report. 
2. What user(s) created the draft invoice on account 130-N0037 and was it part of the 9/28/19 NCSM billing cycle. 
 - If it was part of the Billing cycle, research how it had a different invoice date then the other accounts.  

 
Please reference all the screenshots above and logs as needed.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 16:18

@smasih @amishra @jredfield 

After lengthy phone call and troubleshooting we were able to find the underlining code issue. 

On the account edit screen, the "billing Cycle Drop down" is populating with deactivated billing cycles. They should not be displaying as an option. Please provide an estimate to fix this.

Also, Please provide an estimate to fix the Invoice by site report. 

And Please estimate a fix for not increment the end date on the Review draft invoices report.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 02:21

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 05:44

@kdjstudios @jredfield 

Thank you for the update.

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|X-Small|1|NA|1|NA|0.0|1|


> I ran the Review draft invoice report and saw that this report shows that account had it's invoice at a different time? Also I presume that the report is showing the "create at date" and not the "invoice date" in the invoice date column? Please confirm as we may need to fix the report.

* Yes, Kyle's you are right. As per the current functionality of SAB, the `Invoice Date` column showing the invoice created at date.


> - Look into the Review draft invoices report and determine why the end date range keeps incrementing every time we run the report.

* As per the current functionality of SAB, the code has written in such a way that when every time we click on the go button system automatically update the end date with 1 day.

> - What user(s) created the draft invoice on account 130-N0037 and was it part of the 9/28/19 NCSM billing cycle.

* We looked and found there was no draft invoice present on the account `130-N0037`.

![image](/uploads/59804c6871915dc4cff2f71a32026957/image.png)

>  - If it was part of the Billing cycle, research how it had a different invoice date then the other accounts.

 * As per my best understanding, There are two `Sub Billing Cycle` with the same name but different `Billing Cycle Date`. 
  
     1 - NCSM 1st Cycle Accounts (Last Billed on: 10/29/2018), this billing is deactivated. (Now, someone changed the name with (**(Deactivated) NCSM 1st Cycle Accounts**))  
    2- NCSM 1st Cycle Accounts(Last Billed on: 08/28/2019), this billing cycle is activated.
   * Some Account associated with deactivated billing cycle and some account associated with activated Billing cycle that's why it's showing different billing cycle date on the stage fee section. For more detail check the `Account Details` report.

> On the account edit screen, the "billing Cycle Drop down" is populating with deactivated billing cycles. They should not be displaying as an option. Please provide an estimate to fix this.

* Less than `30 min`

> Also, Please provide an estimate to fix the Invoice by site report.

* Already fixed on the production server.

> And Please estimate a fix for not increment the end date on the Review draft invoices report.   

* Less than `15 min`

Hope this helps

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 11, 2019, 06:01

@amishra @smasih Approved

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 11, 2019, 10:47

@amishra @smasih @jredfield 

Just a quick recap of items on this ticket. 

*  Fix the Invoice by site report - @smasih stated this was just a band-aid fix previously. Was the full fix released?

*  Fix Account edit page - 30 Minutes Approved.

*  Fix End date on Review Draft Invoice Report - 15 min approved.

*  Fix Invoice Date on Review Draft Invoice Report - Needing estimate

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 18, 2019, 04:56

@kdjstudios @jredfield 

> Fix the Invoice by site report - @smasih stated this was just a band-aid fix previously. Was the full fix released?

Yes, this was the full fixed.

> Fix Invoice Date on Review Draft Invoice Report - Needing estimate

* As per my understanding, we need to show the invoice date. right? 
* If yes, then It'll take less than 15 min and we will fix it.  

> * Fix Account edit page - 30 Minutes Approved.

While developing this functionality, one question came to my mind:

* Account has one selected deactivated Billing Cycle and showing deactivated selected billing cycle on the dropdown. 
* But once we will change the functionality to show only active billing cycle on the dropdown list then no longer to show deactivated billing cycle. Now, the Account will show as a selected master billing cycle and then once we will click on the save button accounts will come under the master billing cycle.

So, do we need to show any warning message for users if any accounts have selected a deactivated billing cycle?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 18, 2019, 14:34

@amishra @jredfield 

Thank you for thinking about that, I think we will need a list of all accounts that currently have a deactivated billing cycle selected. We can then have ops update those before we release this fix, that way we do not assign any accounts to the wrong billing cycle.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 19, 2019, 04:52

@kdjstudios @jredfield 

Thank you for the update.

We ran the script and found only 2 accounts of the `Phoenix Synergy` site that have selected the deactivated billing cycle.  **hosted server** has no accounts.

 | Site Name       | Account Number | Account Name        | Billing Cycle                         | Account Status |
|-----------------|----------------|---------------------|---------------------------------------|----------------|
| Phoenix Synergy | 130-N0041      | NCSM Smith & Wesson | (Deactivated) NCSM 1st Cycle Accounts | True           |
| Phoenix Synergy | T-130-N0041    | NCSM Smith & Wesson | (Deactivated) NCSM 1st Cycle Accounts | False          |

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 21, 2019, 24:00

@kdjstudios @jredfield 

* Fix Account edit page - Fixed

* Fix End date on Review Draft Invoice Report - Fixed

* Fix Invoice Date on Review Draft Invoice Report - Fixed(For now on, It'll show the invoice date)
![image](/uploads/28c6eeec5422be32e24753410670bad6/image.png)

We have set up the branch on the merge2 server for your testing. Please test it.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 21, 2019, 11:23

I wrote:

> While the support team was developing a fix to prevent users from selecting a deactivated billing cycle on accounts they found the following two accounts are currently setup to a deactivated Billing cycle. We would like to recommend the site update both accounts to a active billing cycle on the edit account page. Please let us know if there are any questions or concerns. 
> 
> Accounts: 130-N0041 and T-130-N0041

Cheryl wrote:

> I believe these have been corrected. Please let me know if there are any other accounts like this that need to be fixed.


@amishra 

It would appear that the report end date is not being included. Such I created an invoice today and the end date was set to today's date. It should include that invoice, but it is not.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 22, 2019, 24:59

@kdjstudios @jredfield 

Thank you for the update.

* The date range filter works with the invoice generated at(current Time) not invoice date.
* I would like to inform you that the end date should be greater than the current date i.e if you create an invoice today(22-10-2019) then the end date should be (today + 1.day = 23-10-2019). then you are able to see the draft invoices.
 
![image](/uploads/4d6050fc3de6d9140f518c5e29bf7f4e/image.png)

My question is:

* Do we need to change the date range functionality?
* Do we need to add the one column on the report section that will show the invoice generated at.?


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 22, 2019, 09:49

@amishra Are invoices that are created on the start date included? If so, we should do the same for the end date would be my thought.

Adding the Invoice Generated/Created date I think would be a great improvement, which would really help the report out, but I think that should be Gary's call. 

@jredfield  What are your thoughts?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 22, 2019, 23:24

@kdjstudios @jredfield 


> Are invoices that are created on the start date included? If so, we should do the same for the end date would be my thought.

No, the start date does not include.

Let's explain with an example: 

* Invoice Created at or generated at -> 2019-10-23 05:09:33 UTC

Case :

1. When the start date & end date both are same as the invoice generated at-> `No Result`
![image](/uploads/460da51e7c7c24db6698da5b8e908145/image.png)
1. When the start date less than the invoice generated at end date with same as an invoice generated at -> `No result` 
![image](/uploads/a28b69b386225893af8c1567bc79b3b4/image.png)
1. When the start date same as generated at and the end date greater than the invoice generated at-> `Results found`
![image](/uploads/eddc24aa74758e607ca4a84461574e8c/image.png) 

##### In Short, the generated at saving the date with `datetime format`  and we filter with the only date format.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 23, 2019, 11:00

@amishra 

Both start and end dates should be including the results. your example 1 is a perfect scenario, as you want to see only invoices on drafted on one date, but none are being shown due to how it currently is working. May we get an estimate to fix this? Also include in separate estimate adding the Created/drafted at date in a new column please.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 24, 2019, 04:47

@kdjstudios @jredfield 

Thank you for the update.

We have fixed the issue and both dates are including the results. We have set up a branch on the merge2 server for your testing. Please test it.  

###### Please let us know we need to add the column(Generated/Created date) on the draft invoice report. 

![image](/uploads/51c27e62cf4462703a3966a82eacbe7b/image.png)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 24, 2019, 10:07

@amishra 

Thank you for the update. The date filter is working as expected now.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 24, 2019, 10:09

To Gary:

On the Review Draft Invoice Report, the column has an invoice date which can cause some confusion because the draft invoice report is filtering by created/drafted date, not the invoice date. We believe adding an additional "Create Date" column to the report, would help make this report more useful and easier to understand.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 09:18

Gary, deemed that including another date on the report would most likely only add more confusion. Since no users have complained about this yet.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 9, 2019, 03:24

@kdjstudios @jredfield 

We have successfully released on both of the production servers. Please have a look.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 10:23

closed
